File: doubly-linked-list/doubly-linked-list.java
class Node {
    int data;
    Node prev, next;

    Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class DoublyLinkedList {
    private Node head;

    public void insertAtStart(int data) {
        Node newNode = new Node(data);
        if (head != null) {
            newNode.next = head;
            head.prev = newNode;
        }
        head = newNode;
    }

    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
        newNode.prev = temp;
    }

    public void insertAfter(int target, int data) {
        Node temp = head;
        while (temp != null && temp.data != target) temp = temp.next;
        if (temp == null) {
            System.out.println("Target " + target + " not found in the list.");
            return;
        }
        Node newNode = new Node(data);
        newNode.next = temp.next;
        newNode.prev = temp;
        if (temp.next != null) temp.next.prev = newNode;
        temp.next = newNode;
    }

    public void deleteAtStart() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        head = head.next;
        if (head != null) head.prev = null;
    }

    public void deleteAtEnd() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        if (temp.prev != null) temp.prev.next = null;
        else head = null;
    }

    public void deleteSpecific(int target) {
        Node temp = head;
        while (temp != null && temp.data != target) temp = temp.next;
        if (temp == null) {
            System.out.println("Target " + target + " not found in the list.");
            return;
        }
        if (temp.prev != null) temp.prev.next = temp.next;
        else head = temp.next;
        if (temp.next != null) temp.next.prev = temp.prev;
    }

    public void display() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        Node temp = head;
        System.out.print("List elements: ");
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
}

public class DLL {
    public static void main(String[] args) {
        DoublyLinkedList dll = new DoublyLinkedList();
        java.util.Scanner sc = new java.util.Scanner(System.in);

        while (true) {
            System.out.println("1. Insert at Start\n2. Insert at End\n3. Insert After\n4. Delete at Start\n5. Delete at End\n6. Delete Specific\n7. Display\n8. Exit");
            System.out.print("Enter your choice: ");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter data to insert at start: ");
                    dll.insertAtStart(sc.nextInt());
                    break;
                case 2:
                    System.out.print("Enter data to insert at end: ");
                    dll.insertAtEnd(sc.nextInt());
                    break;
                case 3:
                    System.out.print("Enter target after which to insert: ");
                    int target = sc.nextInt();
                    System.out.print("Enter data to insert: ");
                    dll.insertAfter(target, sc.nextInt());
                    break;
                case 4:
                    dll.deleteAtStart();
                    break;
                case 5:
                    dll.deleteAtEnd();
                    break;
                case 6:
                    System.out.print("Enter data to delete: ");
                    dll.deleteSpecific(sc.nextInt());
                    break;
                case 7:
                    dll.display();
                    break;
                case 8:
                    sc.close();
                    System.exit(0);
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

File: doubly-linked-list/doubly-linked-list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_at_start(self, data):
        new_node = Node(data)
        if self.head:
            self.head.prev = new_node
            new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node
        new_node.prev = temp

    def insert_after(self, target, data):
        temp = self.head
        while temp and temp.data != target:
            temp = temp.next
        if not temp:
            print(f"Target {target} not found in the list.")
            return
        new_node = Node(data)
        new_node.next = temp.next
        new_node.prev = temp
        if temp.next:
            temp.next.prev = new_node
        temp.next = new_node

    def delete_at_start(self):
        if not self.head:
            print("List is empty.")
            return
        self.head = self.head.next
        if self.head:
            self.head.prev = None

    def delete_at_end(self):
        if not self.head:
            print("List is empty.")
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        if temp.prev:
            temp.prev.next = None
        else:
            self.head = None

    def delete_specific(self, target):
        temp = self.head
        while temp and temp.data != target:
            temp = temp.next
        if not temp:
            print(f"Target {target} not found in the list.")
            return
        if temp.prev:
            temp.prev.next = temp.next
        else:
            self.head = temp.next
        if temp.next:
            temp.next.prev = temp.prev

    def display(self):
        if not self.head:
            print("List is empty.")
            return
        temp = self.head
        print("List elements: ", end="")
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()

if __name__ == "__main__":
    dll = DoublyLinkedList()

    while True:
        print("1. Insert at Start\n2. Insert at End\n3. Insert After\n4. Delete at Start\n5. Delete at End\n6. Delete Specific\n7. Display\n8. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            ele = int(input("Enter data to insert at start: "))
            dll.insert_at_start(ele)
        elif choice == 2:
            ele = int(input("Enter data to insert at end: "))
            dll.insert_at_end(ele)
        elif choice == 3:
            target = int(input("Enter target after which to insert: "))
            ele = int(input("Enter data to insert: "))
            dll.insert_after(target, ele)
        elif choice == 4:
            dll.delete_at_start()
        elif choice == 5:
            dll.delete_at_end()
        elif choice == 6:
            target = int(input("Enter data to delete: "))
            dll.delete_specific(target)
        elif choice == 7:
            dll.display()
        elif choice == 8:
            break
        else:
            print("Invalid choice.")
